internal/configfsm/fsm.go

```diff
@@ -0,0 +1,315 @@
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case ClusterNotReady:
+		err = fsm.Fire(ClusterNotReadyToClusterNotReady, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case InProgress:
+		err = fsm.Fire(InProgressToInProgress, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case TimedOut:
+		err = fsm.Fire(TimedOutToTimedOut, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case Completed:
+		err = fsm.Fire(CompletedToCompleted, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case OutOfDate:
+		err = fsm.Fire(OutOfDateToOutOfDate, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	}
+	// Update the current state
+	aState, err = fsm.State(ctx)
+	if err != nil {
+		return aState, fmt.Errorf("could not get state, err: %w", err)
+	}
+	return aState, nil
+}
+
+// Initialize the state machine algorithm
+func InitFSM(state string) (fsm *stateless.StateMachine, err error) {
+	fsm = stateless.NewStateMachine(state)
+	fsm.Configure(Start).Permit(StartToMissing, Missing)
+	fsm.Configure(ClusterNotReady).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering ClusterNotReady")
+			if fsmHelper.IsTimedOut() {
+				return fsm.Fire(ClusterNotReadyToTimedOut, fsmHelper)
+			}
+			if fsmHelper.IsClusterReady() {
+				if fsmHelper.IsNonCompliantPolicyInEnforce() || fsmHelper.IsAllPoliciesCompliant() {
+					return fsm.Fire(ClusterNotReadyToInProgress, fsmHelper)
+				} else {
+					return fsm.Fire(ClusterNotReadyToOutOfDate, fsmHelper)
+				}
+			}
+			return nil
+		}).
+		Permit(ClusterNotReadyToMissing, Missing).
+		Permit(ClusterNotReadyToInProgress, InProgress).
+		Permit(ClusterNotReadyToTimedOut, TimedOut).
+		Permit(ClusterNotReadyToOutOfDate, OutOfDate).
+		PermitReentry(ClusterNotReadyToClusterNotReady)
+	fsm.Configure(InProgress).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering InProgress")
+			if fsmHelper.IsResetNonCompliantAt() {
+				fsmHelper.SetResetNonCompliantAtNow()
+			}
+
+			if fsmHelper.IsTimedOut() {
+				return fsm.Fire(InProgressToTimedOut, fsmHelper)
+			}
+			if fsmHelper.IsAllPoliciesCompliant() {
+				return fsm.Fire(InProgressToCompleted, fsmHelper)
+			}
+			if !fsmHelper.IsPoliciesMatched() {
+				return fsm.Fire(InProgressToMissing, fsmHelper)
+			}
+			if !fsmHelper.IsClusterReady() {
+				return fsm.Fire(InProgressToClusterNotReady, fsmHelper)
+			}
+			if !fsmHelper.IsNonCompliantPolicyInEnforce() && !fsmHelper.IsAllPoliciesCompliant() {
+				return fsm.Fire(InProgressToOutOfDate, fsmHelper)
+			}
+			return nil
```

Comment:
Nil in this context means no error, so there would be nothing to do. If any of the Fire function return an error, The RunFSM function returns an error. This could happen if the state that is fired is producing a non permitted transition.

Reply:
right so the FSM will remain in the current state until it fires again and will repeat that until one of the preconditions matches to move it to another state?

Reply:
yes, correct. The RunFSM runs the state transitions to completion from the current state until it is called again. When it reaches "return nil", the RunFSM exits.